stan/math/opencl/copy.hpp

```diff
@@ -29,11 +31,34 @@ namespace math {
  * matrices do not have matching dimensions
  */
 template <int R, int C>
-void copy(matrix_cl& dst, const Eigen::Matrix<double, R, C>& src) {
+inline void copy(matrix_cl& dst, const Eigen::Matrix<double, R, C>& src) {
   check_size_match("copy (Eigen -> (OpenCL))", "src.rows()", src.rows(),
                    "dst.rows()", dst.rows());
   check_size_match("copy (Eigen -> (OpenCL))", "src.cols()", src.cols(),
                    "dst.cols()", dst.cols());
+  if (src.size() > 0) {
+    internal::cache_copy(dst.buffer(), src);
+  }
+}
+
+/**
+ * Copies the source var type Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param dst destination matrix on the GPU
+ * @param src source Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ */
+template <int R, int C>
+inline void copy(matrix_cl& dst, const Eigen::Matrix<var, R, C>& src) {
```

Comment:
Can this function just call the overload above with `value_of(src)`?

Reply:
The above one uses cache, we dont want to use cache on var matrices, hence the specializations.

Reply:
it's actually fine if the matrix returned by `value_of` gets the cl::Buffer put on it, no? :P

Reply:
@steveBronder you tested this one right?, do we lose performance if we do this?

Reply:
tagging @SteveBronder again - did we lose performance if we just called the caching copy with the `value_of` the var matrix? if not we can just farm it out, if so we can merge the packed transfer branch into this and update it and maybe get it in for tomorrow

Reply:
Never mind! there's the creation of an additional cache buffer I didn't realize.